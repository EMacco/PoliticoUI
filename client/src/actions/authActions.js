import axios from 'axios';
import setAuthToken from '../utils/setAuthToken';
import {
  GET_ERRORS,
  SET_CURRENT_USER,
  CLEAR_CURRENT_PROFILE,
  CLEAR_DASHBOARD,
  CLEAR_PARTIES,
  CLEAR_OFFICES
} from '../actions/types';

// Set logged in user
export const setCurrentUser = user => {
  return {
    type: SET_CURRENT_USER,
    payload: user
  };
};

// Register user
export const registerUser = userData => dispatch => {
  axios
    .post('/auth/signup', userData)
    .then(res => {
      const response = res.data.data[0];

      // Set token in local storage
      const { token } = response;
      localStorage.setItem('jwtToken', token);

      // Set token to auth header
      setAuthToken(token);

      // Decode to get user data
      const { user } = response;

      // Set current user
      dispatch(setCurrentUser(user));
    })
    .catch(err =>
      dispatch({
        type: GET_ERRORS,
        payload: { global: err.response.data.error }
      })
    );
};

export const loginUser = userData => dispatch => {
  axios
    .post('/auth/login', userData)
    .then(res => {
      const response = res.data.data[0];

      // Set token in local storage
      const { token } = response;
      localStorage.setItem('jwtToken', token);

      // Set token to auth header
      setAuthToken(token);

      // Decode to get user data
      const { user } = response;

      // Set current user
      dispatch(setCurrentUser(user));
    })
    .catch(err => {
      dispatch({ type: GET_ERRORS, payload: { global: err.response.data.error } });
    });
};

// Log User out
export const logoutUser = history => dispatch => {
  localStorage.removeItem('jwtToken');
  setAuthToken(false);
  dispatch(setCurrentUser({}));
  dispatch({ type: CLEAR_CURRENT_PROFILE });
  dispatch({ type: CLEAR_DASHBOARD });
  dispatch({ type: CLEAR_PARTIES });
  dispatch({ type: CLEAR_OFFICES });
  if (history) history.push('/');
  else window.location.href = '/';
};
