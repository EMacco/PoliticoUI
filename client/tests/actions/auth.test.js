import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { loginUser, registerUser, logoutUser } from '@actions/authActions';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const store = mockStore({
  auth: {},
  dashboard: {},
  parties: {},
  offices: {}
});

const newUser = {
  firstname: 'John',
  lastname: 'Doe',
  email: 'testing@testing.com',
  phone: '08088293829',
  password: 'password'
};

const history = { push: jest.fn() };

describe('Auth actions', () => {
  it('should dispatch SET_CURRENT_USER on successful registration', async () => {
    store.dispatch(registerUser(newUser));
    await expect(store.getActions()).toMatchSnapshot();
  });

  it('should dispatch SET_CURRENT_USER, CLEAR_CURRENT_PROFILE, CLEAR_DASHBOARD, CLEAR_PARTIES, and CLEAR_OFFICES', () => {
    store.dispatch(logoutUser(history));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should dispatch SET_CURRENT_USER on successful login', async () => {
    await store.dispatch(loginUser({ email: 'testing@testing.com', password: 'password' }));
    expect(store.getActions()).toMatchSnapshot();
  });
});
