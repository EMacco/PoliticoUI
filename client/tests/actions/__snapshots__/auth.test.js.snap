// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth actions should dispatch SET_CURRENT_USER on successful login 1`] = `
Array [
  Object {
    "payload": Object {
      "email": "testing@testing.com",
      "firstname": "John",
      "id": 1,
      "lastname": "Doe",
      "password": "password",
      "phone": "08088293829",
    },
    "type": "SET_CURRENT_USER",
  },
  Object {
    "payload": Object {},
    "type": "SET_CURRENT_USER",
  },
  Object {
    "type": "CLEAR_CURRENT_PROFILE",
  },
  Object {
    "type": "CLEAR_DASHBOARD",
  },
  Object {
    "type": "CLEAR_PARTIES",
  },
  Object {
    "type": "CLEAR_OFFICES",
  },
  Object {
    "payload": Object {
      "email": "testing@testing.com",
      "firstname": "John",
      "id": 1,
      "lastname": "Doe",
      "password": "password",
      "phone": "08088293829",
    },
    "type": "SET_CURRENT_USER",
  },
]
`;

exports[`Auth actions should dispatch SET_CURRENT_USER on successful registration 1`] = `Array []`;

exports[`Auth actions should dispatch SET_CURRENT_USER, CLEAR_CURRENT_PROFILE, CLEAR_DASHBOARD, CLEAR_PARTIES, and CLEAR_OFFICES 1`] = `
Array [
  Object {
    "payload": Object {
      "email": "testing@testing.com",
      "firstname": "John",
      "id": 1,
      "lastname": "Doe",
      "password": "password",
      "phone": "08088293829",
    },
    "type": "SET_CURRENT_USER",
  },
  Object {
    "payload": Object {},
    "type": "SET_CURRENT_USER",
  },
  Object {
    "type": "CLEAR_CURRENT_PROFILE",
  },
  Object {
    "type": "CLEAR_DASHBOARD",
  },
  Object {
    "type": "CLEAR_PARTIES",
  },
  Object {
    "type": "CLEAR_OFFICES",
  },
]
`;
